### DISCUSSION TITLE
ESP32-C3 failed to create SPI device

### DISCUSSION THREAD

**Discussion by JohnMasen** [nanoframework/Home#1220] [Q&A]
the following code tries to create SPI1 device, it will crash on my ESP32-C3 device.
```
        public static void Main()
        {
            SpiConnectionSettings settings = new(1, -1)
            {
                ClockFrequency = 2_400_000,
                Mode = SpiMode.Mode0,
                DataBitLength = 8
            };
            SpiDevice spi1 = SpiDevice.Create(settings);
        }
```

try to get the clock frequency for SPI1 will also crash

`SpiDevice.GetBusInfo(1).MinClockFrequency`

I did some research, seems C3 chips has different SPI layout than ESP32.  ref:[https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/spi_master.html#:~:text=Overview%20of%20ESP32%E2%80%99s%20SPI%20peripherals%20ESP32%20integrates%204,arbiter%20to%20determine%20which%20can%20access%20the%20bus.](url)

Not sure if my code is not correct or missing some something, many thanks for your help.

**Comment by Ellerbach** [comment]
Did you setup the pins before opening SPI1?

**Reply by Ellerbach** [reply]
The default pin setup for the ESP32-C3 setup in the firmware are those ones: https://github.com/nanoframework/nf-interpreter/blob/main/targets/ESP32/_common/ESP32_C3_DeviceMapping.cpp#L13
(note that the link you've shared doesn't seem to work for me)

**Reply by Ellerbach** [reply]
And the default SPI pins for the C3 is here in the IDF code: https://github.com/espressif/esp-idf/blob/master/components/soc/esp32s2/include/soc/spi_pins.h

**Reply by JohnMasen** [reply]
Hi Ellerbach,
thanks for your support. The _spi_pins.h_ you're using is for esp32s2 chips. I found the corresponding file under C3 folder at [https://github.com/espressif/esp-idf/blob/master/components/soc/esp32c3/include/soc/spi_pins.h](url)

After modified my code to align with pin definitions, it still crashes will same exception.
```
public static void Main()
        {
            Configuration.SetPinFunction(16, DeviceFunction.SPI1_MOSI);
            Configuration.SetPinFunction(17, DeviceFunction.SPI1_MISO);
            Configuration.SetPinFunction(15, DeviceFunction.SPI1_CLOCK);
            SpiConnectionSettings settings = new(1, 14)
            {
                ClockFrequency = 2_400_000,
                Mode = SpiMode.Mode0,
                DataBitLength = 8
            };
            SpiDevice spi1 = SpiDevice.Create(settings);
      }
```

**Reply by JohnMasen** [reply]
I found the [SPI Sample](https://github.com/nanoframework/Samples/blob/main/samples/SPI/Program.cs) is creating spidevice without call `SetPinFunction`.  I guess SPI1 device is defined in the DeviceMapping.cpp, it should be able to created without set the pin functions manually, unless I want change the pin definition for SPI1.
